contributors/design-proposals/scheduling/scheduling-framework.md

```diff
@@ -0,0 +1,406 @@
+cycle** to schedule the pod. At the end of the cycle scheduler determines
+whether the pod is schedulable or not. If the pod is not schedulable, its status
+is updated and goes back to the scheduling queue. If the pod is schedulable (one
+or more nodes are found that can run the Pod), scoring process is started. The
+scoring process finds the best node to run the Pod.  a bind go routine is
+started to bind the pod.  
+The above process is the same as what Kubernetes scheduler v1 does. Some of the
+essential features of scheduler v1, such as leader election, will also be
+transferred to the scheduling framework.  
+In the rest of this section we describe how various plugins are used to enrich
+this basic workflow. In this section, we focus on in-process plugins.
+Out-of-process plugins are discussed later in the doc.
+
+## Communication and statefulness of plugins
+
+The scheduling framework provides a library that plugins can use to pass
+information to other plugins. This library keeps a map from keys of type string
+to opaque pointers of type interface{}. A write operation takes a key and a
+pointer and stores the opaque pointer in the map with the given key. Other
+plugins can provide the key and receive the opaque pointer. Multiple plugins can
+share the state or communicate via this mechanism.  
+The saved state is preserved only during a single scheduling cycle. At the end
+of a scheduling cycle, this map is destructed. So, plugins cannot keep shared
+state across multiple scheduling cycle. They can, however, update the scheduler
+cache via the provided interface of the cache. The cache interface allows
+limited state preservation across multiple scheduling cycle.  
+It is worth noting that plugins are assumed to be **trusted**. Scheduler does
+not prevent one plugin from accessing or modifying another plugin's state.
+
+## Plugin registration
+
+Plugin registration is done by providing an extension point and a function that
+should be called at that extension point. This step will be something like:
+
+<table>
+<thead>
+<tr>
+<th><p><pre>
+register("pre-filter", plugin.foo)
+</pre></p>
+
+</th>
+</tr>
+</thead>
+<tbody>
+</tbody>
+</table>
+
+The details of the function signature will be provided later.
+
+## Extension points
+
+The following picture shows the scheduling cycle of a Pod and the extension
+points that the scheduling framework exposes. In this picture "Filter" is
+equivalent to "Predicate" in scheduler v1 and "Scoring" is equivalent to
+"Priority function". Plugins are go functions. They are registered to be called
+at one of these extension points. They are called by the framework in the same
+order they are registered for each extension point.  
+In the following sections we describe each extension point in the same order
+they are called in a schedule cycle.
+
+![image](images/scheduling-framework-extensions.png)
+
+### Scheduling queue sort
+
+These plugins indicate how Pods should be sorted in the scheduling queue. A
```

Comment:
Is there a mechanism which ensures that Pods further in the queue will not be starved by Pods with higher priority.
E.g. can we defend ourselves from situation where we pick a first not-scheduled Pod from sorted queue. Pass it through scheduling steps but at some point we decide to Reject it. IIUC it now gets back to the queue. And if it ends up (after resorting of queue) to be the first one in the queue again, the situation may repeat (possibly indefinitely).

Or maybe the resorting only happens after all the pods from the queue have been considered? I did not get that from the doc.

Reply:
same doubt here.

Reply:
Our current scheduler (with Pod Priority and Preemption enabled) has a smarter scheduling queue that does not retry scheduling a Pod unconditionally after it has determined unschedulable. The scheduler stores a Pod in a separate structure and brings it back to the scheduling queue only if there is an event in the cluster that could make the Pod schedulable, for example, another Pod is terminated, .... This will effectively prevent a high priority Pod from hugging the front of the scheduling queue in most practical scenarios. The starvation scenario could still happen if the cluster has very high churn and many pods are terminated every second. In that case a high priority unschedulable pod may be retried many times per second as well. We are building some backoff mechanism to address this unlikely scenario as well. A similar approach should be used here too.